onnxruntime/core/mlas/lib/jblas_gemm.cpp

```diff
@@ -0,0 +1,525 @@
+                            DataParams[i].ldc, WorkSpace, &orth
+                        );
+                    } else if (NTile == tAVX_VNNI::NTILE && _cd->AVX_VNNI()) {
+                        JblasSQ4GemmCompInt8<tAVX_VNNI>(
+                            M, N, K, DataParams[i].A, DataParams[i].lda, kptr, DataParams[i].C,
+                            DataParams[i].ldc, WorkSpace, &orth
+                        );
+                    }
+                }
+                if (CType == uint32_t(gemm::CompType::COMP_INT8_SS_INT32)) {
+                    if (NTile == tAMX_INT8_SS::NTILE && _cd->AMX_INT8()) {
+                        JblasSQ4GemmCompInt8<tAMX_INT8_SS>(
+                            M, N, K, DataParams[i].A, DataParams[i].lda, kptr, DataParams[i].C,
+                            DataParams[i].ldc, WorkSpace, &orth
+                        );
+                    }
+                }
+            }
+            delete ptr;
+        } else {
+            processed = false;
+            break;
+        }
+    }
+    return processed;
+}
+
+template <class GemmCore_T>
+static size_t
+JblasSQ4GemmCompF32WorkspaceSize(
+    const int M,
+    const int N,
+    const int K,
+    const float* A,
+    const int lda,
+    jblas::storage::gemm::StorageWeightKBlockS4* B,
+    float* C,
+    const int ldc
+)
+{
+    if (M <= 32) {
+        using Launcher = tLauncher_Fp32_S4_F32F32<GemmCore_T>;
+        static Launcher kernel;
+        if (B->mIsAsym) {
+            auto reduceA = kernel.mProA.createStorage(M, K, B->mBlockSize);
+            return reduceA.mSize;
+        }
+        return 0;
+    } else {
+        using Launcher = jblas::wrapper::gemm::LauncherBase<
+            GemmCore_T::ISA, GemmCore_T, jblas::prologue_a::gemm::ActivationBase,
+            jblas::prologue_b::gemm::WeightKBlockS4, jblas::epilogue::gemm::AccumulatorWriteBackFp32>;
+        static Launcher kernel;
+        return 0;
+    }
+    return 0;
+}
+
+template <class GemmCore_T>
+static size_t
+JblasSQ4GemmCompInt8WorkspaceSize(
+    const int M,
+    const int N,
+    const int K,
+    const float* A,
+    const int lda,
+    jblas::storage::gemm::StorageWeightKBlockS4* B,
+    float* C,
+    const int ldc
+)
+{
+    using Parallel = jblas::parallel::gemm::SchedulerKBlock<GemmCore_T>;
+    using Launcher = tLauncher_Int8_S4_F32F32<GemmCore_T>;
+
+    static Launcher kernel;
+    auto quanA = kernel.mProA.createStorage(M, K, B->mBlockSize, B->mIsAsym);
+    return quanA.mSize;
+}
+
+size_t
+JblasSQ4GemmBatchWorkspaceSize(
+    const size_t M,
+    const size_t N,
+    const size_t K,
+    const size_t BatchN,
+    const MLAS_SQNBITS_GEMM_DATA_PACKED_PARAMS* DataParams
+)
+{
+    GetCPUDevice();
+    size_t size = 0;
+    for (size_t i = 0; i < BatchN; i++) {
+        auto ptr =
+            jblas::storage::gemm::PackedWeightParser::deserialBuffer(const_cast<void*>(DataParams[i].B));
```

Comment:
general: can we avoid const_cast?

Reply:
I can do that. May I ask why? I don't feel any difference.

Reply:
https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es50-dont-cast-away-const

to be clear, by "avoid const_cast" I meant "not cast away const" and not "replace with a C-style cast".

Reply:
removed